File: SlidingWindows.java
import java.util.HashMap;

public class SlidingWindows {

    public int lengthOfLongestSubstring(String s) {

        int left = 0;
        int right;
        HashMap<Character, Integer> hashMap = new HashMap<>();
        int maxSubst = Integer.MIN_VALUE;

        if (s.isEmpty()) return 0;
        if (s.length() == 1) return 1;

        for (right=0; right<s.length(); right++){

            if (!hashMap.containsKey(s.charAt(right))) {
                hashMap.put(s.charAt(right), right);

            } else {

                if (hashMap.size() >= maxSubst){
                    maxSubst = hashMap.size();
                }

                while (s.charAt(left) != s.charAt(right)){
                    hashMap.remove(s.charAt(left));
                    left++;
                }
                hashMap.remove(s.charAt(left));
                hashMap.put(s.charAt(right), right);
                left++;
            }
        }
        return Math.max(maxSubst, hashMap.size());
    }

    public static void main(String[] args) {
        SlidingWindows SlidingWindows = new SlidingWindows();
        String test = "pwwkew";
        System.out.println(SlidingWindows.lengthOfLongestSubstring(test));
    }
}

File: SlidingWindows.kt
class SlidingWindows {

    fun lengthOfLongestSubstring(s: String): Int {

        var left = 0
        var maxSubst = Int.MIN_VALUE
        val hashMap = HashMap<Char, Int>()

        for (right in 0 until s.length){

            if (!hashMap.containsKey(s.get(right))){
                hashMap.put(s[right], right)
            } else {

                if (hashMap.size >= maxSubst) {
                    maxSubst = hashMap.size
                }

                while (s[left] != s[right]){

                    hashMap.remove(s[left])
                    left++
                }
                hashMap.remove(s[left])
                hashMap.put(s[right], right)
                left++;

            }
        }
        return Math.max(hashMap.size, maxSubst)
    }

    // 1004. Max Consecutive Ones III
    fun longestOnes(nums: IntArray, k: Int): Int {

        var right = 0
        var left = 0
        var maxCount = 0
        var hashmap = 0

        while (right < nums.size){
            if (nums[right] == 0){
                if (hashmap < k){
                    hashmap += 1
                    maxCount = Math.max(maxCount, right - left)

                }else{
                    maxCount = Math.max(maxCount, right - left)
                    while (hashmap >= k){
                        if (nums[left] == 0){
                            hashmap -= 1
                            left += 1
                        }else{
                            left += 1
                        }
                    }
                    hashmap += 1
                }
            }else{
                maxCount = Math.max(maxCount, right - left)
            }
            right += 1
        }

        return Math.max(maxCount, right - left)
    }


}

fun main(args : Array<String>){

    val testClass = SlidingWindows()
    val testString = "pwwkew"
    val testInput = intArrayOf(0,0,1,1,0,0,1,1,1,0,1,1,0,0,0,1,1,1,1)
    println(testClass.longestOnes(testInput, 3))

}